### ISSUE TITLE
support for decibels?

### ISSUE THREAD

**Issue by t-kalinowski** [r-quantities/units#176]
I'm not entirely sure how this should work. 

Short of a holistic solution, a first step might be sensible printing if you recreate the unit manually (so the `lg(1)` below might instead be `dB`
``` r
library(units)
#> udunits system database from /usr/share/udunits
signal_power <- set_units(10, W)
noise_power <- set_units(2, W)

snr <- 10 * log(signal_power / noise_power, base = 10)
snr
#> 6.9897 [lg(1)]
```

(as a side note, it would also be good to add support for `log10` and `log2` at some point, so the above `log(..., base = 10)` wouldn't be necessary.

<sup>Created on 2018-09-26 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

**Comment by edzer** [comment]
`udunits` apparently knows about the `dB` [prefix](https://en.wikipedia.org/wiki/Decibel):
```r
> set_units(1:3, dBm)
Units: [dBm]
[1] 1 2 3
> set_units(1:3, dBV)
Units: [dBV]
[1] 1 2 3
```
I guess your example should result in `[dBW]`?

**Comment by Enchufa2** [comment]
It is not a prefix, but a unit:

```r
units:::R_ut_format(units:::R_ut_parse("dBm"))
#> [1] "0.1 lg(re 0.001 W)"
```

means 0.1 log referenced to 1 mW.

**Comment by t-kalinowski** [comment]
Yeah, dBm is close, but not entirely applicable for dB. dB is really a ratio on a log scale. dBm is a special case of dB with specific known value in the denominator (1 milliwatt). I don't think it would make sense to convert snr (signal-to-noise ratio) to dBm in this case.

**Comment by Enchufa2** [comment]
One way to get dB:

```r
library(units)
#> udunits system database from /usr/share/udunits
install_conversion_constant("lg(re 1)", "dB", 1)
signal_power <- set_units(10, W)
noise_power <- set_units(2, W)
set_units(signal_power / noise_power, dB)
#> 0.69897 [dB]
```

**Comment by Enchufa2** [comment]
@edzer: Perhaps `units` should define that conversion automatically? And we should fix the `log` operation, because `lg(1)` means nothing for udunits:

```r
units:::R_ut_format(units:::R_ut_parse("lg(1)"))
#> Error in units:::R_ut_parse("lg(1)") : 
#>  Error in function R_ut_parse: string unit representation contains unknown word
```

The correct syntax is `lg(re 1)`.

**Comment by Enchufa2** [comment]
See #177 for a PR addressing these issues. This initial implementation fixes the `log` operation, but there are some things that remain to be discussed. Specifically, I was thinking about adding the convenience definition `install_conversion_constant("lg(re 1)", "B", 1)` in `.onLoad`, then manage special cases of interest in `log`:

- If the input unit is W and `base=10`, then convert the output to dBW (instead of `lg(re 1 W)`).
- If the input unit is mW and `base=10`, then convert the output to dBm (instead of `lg(re 0.001 W)`).
- If the input unit is unitless and `base=10`, then convert the output to dB (instead of `lg(re 1)`).

Also, please, @edzer, commit the latest changes to `master` (the ones you made to submit the latest release).

**Comment by edzer** [comment]
Thumbs up meant I committed my last changes. Would be good if @t-kalinowski reviewed this, dB's are a bit remote for me (although I appreciate music, occasionally loud).

**Comment by t-kalinowski** [comment]
Thanks @Enchufa2, this is good progress. I like that log2 and log10 work correctly now.

Re your proposed additional functionality to `log()`: I don't think it's right to do automatic conversion of `log(W)` and `log(V)` into `dBW` and `dBV` (not to be confused with `dBv`!). Mathematically, they're not the same operation; if you take log of a W, it's not dBW because in the actual dBW the units dropped out to `1` before the log was taken (same for dBV). 

For the more general case of `dB`, I'm on the fence but can be persuaded either way. Initially I felt unsure about installing the convenience conversions to `B` and `bel`, since in practice I've only ever seen dB, but after seeing BW and BV in the default database, I think it's ok to add an entry `B`. However, doing automatic converstion in `log()` to dB feels a a step too far and becomes a little too magical. I think that type of conversion should be explicit. (especially because the `deci-` prefix changes the numeric by a factor of 10).

Even with explicit only conversion, I imagine the behavior might be a little surprising to users sometimes, because there might be a conversion where they don't expect it. This is probably something that' can't be avoided. Maybe this needs a note in the documentation.
```R
> install_conversion_constant("0.1 lg(re 1)", "dB", 1)
> 5 %>% set_units(dB)
5 [dB]
> 5 %>% set_units() %>% set_units(dB)
6.9897 [dB]
```

I'm hesitant to work too much on this at the R level because really this type of logic belongs upstream in the C library, and it seems like there has been some (discussion already)[https://github.com/Unidata/UDUNITS-2/issues/33] (unfortunately without much progress)

So, in summary:
  :+1: for automatic addition of B in `.onLoad()`, defined as `lg(re 1)`. If you'd prefer to make the convenience more narrow, I think going with dB == `0.1 lg(re 1)` makes a lot of sense as well. In either case, people should be able to shut this off through some global option (doing this in `units_options()` will be tricky, because that's will only be available after `.onLoad()` has been called. You may have to go with a global `options`. Also, people might have made additions to their existing XML database, we will need to check that those aren't masked inadvertently with this addition)
:-1:  for automatic conversion inside `log()` of `dBW` or `dBV`
:question: for automatic converstion of `log (re 1)` to `dB` (leaning towards no)

**Comment by Enchufa2** [comment]
Thanks for the discussion. Some comments inline:

> Thanks @Enchufa2, this is good progress. I like that log2 and log10 work correctly now.
> 
> Re your proposed additional functionality to `log()`: I don't think it's right to do automatic conversion of `log(W)` and `log(V)` into `dBW` and `dBV` (not to be confused with `dBv`!). Mathematically, they're not the same operation; if you take log of a W, it's not dBW because in the actual dBW the units dropped out to `1` before the log was taken (same for dBV).

I agree to some extent. The BIPM does not permit the use of dBW, dBV and the like alongside SI units (although, in practice, their use is widespread, and that's why they are included in udunits2). I believe this is because, mathematically, these units are not well-defined. If you *really* divide by the same unit and then take the logarithm of a unitless quantity, then the result should be dB, not dBthings. Therefore, IMO, the common definition

![image](https://user-images.githubusercontent.com/4542928/46280808-340a5680-c56d-11e8-9d0e-bea2a5090d0c.png)

should be, instead,

![image](https://user-images.githubusercontent.com/4542928/46280865-556b4280-c56d-11e8-804e-dffefb0cd3b3.png)

In fact, this is the udunits2 approach:

```r
log10(set_units(5.5, W))
#> 0.7403627 [lg(re 1 W)]
```

which means *log10 referenced to 1 W*. My point is that `BW` is just an alias for `lg(re 1 W)`:

```r
set_units(log10(set_units(5.5, W)), BW)
#> 0.7403627 [BW]
```

So there's no conversion here, we are just using an alias when available. That could be one way to go, and it could be optional. The thing is that `BW` is never used, and it's probably hard to recognise (and so is `lg(re 1 W)`), and that's why I was proposing to convert to `dBW`.

> For the more general case of `dB`, I'm on the fence but can be persuaded either way. Initially I felt unsure about installing the convenience conversions to `B` and `bel`, since in practice I've only ever seen dB, but after seeing BW and BV in the default database, I think it's ok to add an entry `B`. However, doing automatic converstion in `log()` to dB feels a a step too far and becomes a little too magical. I think that type of conversion should be explicit. (especially because the `deci-` prefix changes the numeric by a factor of 10).

The case of `dB` is the same: `B` would be an alias for `lg(re 1)`, and the *magic* is the same: take out a factor of 10 from the unit (with the `d` prefix) and add it to the numerical value. Thus, IMO, whatever we decide to do for `lg(re 1 W)` and the like, we should do the same for `lg(re 1)`. 

> Even with explicit only conversion, I imagine the behavior might be a little surprising to users sometimes, because there might be a conversion where they don't expect it. This is probably something that' can't be avoided. Maybe this needs a note in the documentation.
> 
> ```r
> > install_conversion_constant("0.1 lg(re 1)", "dB", 1)
> > 5 %>% set_units(dB)
> 5 [dB]
> > 5 %>% set_units() %>% set_units(dB)
> 6.9897 [dB]
> ```

It should *not* be surprising. This is the same behaviour as always: `set_units` on a numeric value sets up a `units` object; `set_units` on a `units` object performs conversion (if units are compatible). So I don't think that any further explanation is needed.

> I'm hesitant to work too much on this at the R level because really this type of logic belongs upstream in the C library, and it seems like there has been some [discussion already](https://github.com/Unidata/UDUNITS-2/issues/33) (unfortunately without much progress)

As said before, the upstream library already took a decision, that is to set the unit, whichever it is, as the reference for the logarithm. So in summary, my proposal is just

1. to recognise and use the available aliases when `log10` is used, and
2. to move one order of magnitude from the unit to the value, because dBW are widely used and BW are not. I don't have a strong opinion though.

**Comment by Enchufa2** [comment]
#177 merged! Closing this for now. We can revisit automatic conversion in the future if users request it. :)
